homeassistant/components/freebox/device_tracker.py

```diff
@@ -1,65 +1,155 @@
-    """Validate the configuration and return a Freebox scanner."""
-    scanner = FreeboxDeviceScanner(hass.data[DATA_FREEBOX])
-    await scanner.async_connect()
-    return scanner if scanner.success_init else None
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up device tracker for Freebox component."""
+    router = hass.data[DOMAIN][entry.unique_id]
+    tracked = {}
 
+    @callback
+    def update_router():
+        """Update the values of the router."""
+        add_entities(router, async_add_entities, tracked)
 
-Device = namedtuple("Device", ["id", "name", "ip"])
-
-
-def _build_device(device_dict):
-    return Device(
-        device_dict["l2ident"]["id"],
-        device_dict["primary_name"],
-        device_dict["l3connectivities"][0]["addr"],
+    router.listeners.append(
+        async_dispatcher_connect(hass, router.signal_device_new, update_router)
     )
 
-
-class FreeboxDeviceScanner(DeviceScanner):
-    """Queries the Freebox device."""
-
-    def __init__(self, fbx):
-        """Initialize the scanner."""
-        self.last_results = {}
-        self.success_init = False
-        self.connection = fbx
-
-    async def async_connect(self):
-        """Initialize connection to the router."""
-        # Test the router is accessible.
-        data = await self.connection.lan.get_hosts_list()
-        self.success_init = data is not None
-
-    async def async_scan_devices(self):
-        """Scan for new devices and return a list with found device IDs."""
-        await self.async_update_info()
-        return [device.id for device in self.last_results]
-
-    async def get_device_name(self, device):
-        """Return the name of the given device or None if we don't know."""
-        name = next(
-            (result.name for result in self.last_results if result.id == device), None
+    update_router()
+
+
+@callback
+def add_entities(router, async_add_entities, tracked):
+    """Add new tracker entities from the router."""
+    new_tracked = []
+
+    for mac, device in router.devices.items():
+        if mac in tracked:
+            continue
+
+        tracked[mac] = FreeboxDevice(router, device)
+        new_tracked.append(tracked[mac])
+
+    if new_tracked:
+        async_add_entities(new_tracked)
+
+
+class FreeboxDevice(ScannerEntity):
+    """Representation of a Freebox device."""
+
+    def __init__(self, router: FreeboxRouter, device: Dict[str, any]):
+        """Initialize a Freebox device."""
+        self._router = router
+        self._name = device["primary_name"].strip() or DEFAULT_DEVICE_NAME
+        self._mac = device["l2ident"]["id"]
+        self._manufacturer = device["vendor_name"]
+        self._icon = icon_for_freebox_device(device)
+        self._active = False
+        self._attrs = {}
+
+        self._unsub_dispatcher = None
+        _LOGGER.error("ADDED_DEVICE : %s", self.name)
+
+        self.update()
```

Comment:
Remove this call and call `async_add_entities` with second argument `True` instead.